File: editor/en/[1992]Find All Groups of Farmland.java
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    private int minRow;
    private int minCol;
    private int maxRow;
    private int maxCol;

    public int[][] findFarmland(int[][] land) {
        List<int[]> farmlands = new ArrayList<>();
        int m = land.length;
        int n = land[0].length;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (land[i][j] == 1) {
                    minRow = i;
                    maxRow = i;
                    minCol = j;
                    maxCol = j;
                    dfs(land, i, j);
                    farmlands.add(new int[]{minRow, minCol, maxRow, maxCol});
                }
            }
        }

        return farmlands.toArray(new int[0][]);
    }

    private void dfs(int[][] land, int row, int col) {
        if (row < 0 || row >= land.length || col < 0 || col >= land[0].length || land[row][col] != 1) {
            return;
        }

        land[row][col] = 0;
        minRow = Math.min(minRow, row);
        maxRow = Math.max(maxRow, row);
        minCol = Math.min(minCol, col);
        maxCol = Math.max(maxCol, col);
        dfs(land, row - 1, col);
        dfs(land, row + 1, col);
        dfs(land, row, col - 1);
        dfs(land, row, col + 1);
    }
}

//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[1992]Find All Groups of Farmland.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def findFarmland(self, land: List[List[int]]) -> List[List[int]]:
        m = len(land)
        n = len(land[0])
        farmlands = []

        def dfs(i, j):
            nonlocal min_row, min_col, max_row, max_col
            if i < 0 or i >= m or j < 0 or j >= n or land[i][j] != 1:
                return

            land[i][j] = 0
            min_row = min(min_row, i)
            min_col = min(min_col, j)
            max_row = max(max_row, i)
            max_col = max(max_col, j)

            dfs(i - 1, j)
            dfs(i + 1, j)
            dfs(i, j - 1)
            dfs(i, j + 1)

        for i in range(m):
            for j in range(n):
                if land[i][j] == 1:
                    min_row, min_col, max_row, max_col = i, j, i, j
                    dfs(i, j)
                    farmlands.append([min_row, min_col, max_row, max_col])

        return farmlands

# leetcode submit region end(Prohibit modification and deletion)
